### ISSUE TITLE
Feature request: autosaves

### ISSUE THREAD

**Issue by rr-** [LostArtefacts/TRX#600]
#### Rationale
- lots of streamers cussing at the game for saving too long ago. This happens on almost every stream.
- reducing the difficulty. it's convenient in games like Celeste how the characters spawn right after the death.

#### Rough implementation proposal
Save the game every X seconds.

#### Pitfall
Suppose X=10:
- T=0. Game starts / is loaded.
- T=5. Player starts running.
- T=9. Player jumps.
- T=10. Autosave.
- T=11. Lara falls to her death.

Player loads an autosave and Lara is already midair, falling to her death.

#### Detailed implementation proposal that should solve this
- have a special save slot at the top of the save table called AUTOSAVE. Have it only appear if there is an autosave.
- every X seconds, take an in-memory game snapshot using the current savegame facilities. No files involved.
- every time we make such an "autosave", take the previous autosave from X seconds ago, and persist that to HDD.
- when loading an autosave, we're loading that physical savegame file persisted above.

This double buffer helps with the situation above:

- T=0. Game starts.
- T=5. Player starts running.
- T=9. Player jumps.
- T=10. Autosave trigger. But there's no previous autosave so nothing gets saved in HDD, only remembered in the game memory.
- T=11. Lara falls to her death.

Player has to start the game again from scratch.

Scenario 2:
- T=0. Game Starts.
- T=5. Player starts running.
- T=9. Player collects a medpack. They're standing on a rock.
- T=10. Autosave trigger. Remember that the player has a medpack, where they are etc.
- T=19. Player jumps.
- T=20. Autosave trigger. We put into an autosave file that the player was standing on a rock from T=10, and had just picked up a medpack. We take a snapshot how they're falling to their death now.
- T=21. Lara dies.

Player can load autosave from when Lara was standing on a rock.

**Comment by oziphantom** [comment]
I don't see how having two and then the last one being loaded. You could die, then load, then end up in a good spot on the second one, then they die, it loads the first one and they die again. If it resets the timer it might work, but I feel that it is limiting the issue rather than avoiding it outright. 
When we do death reposition in games, we make some "safe position" rules. 
I.e we record the last position they where standing on, haven't taken damage for 10 seconds, 

I feel that is should be mostly trivial for us to do the same in TR1. 
Make sure Lara is on flat solid ground, so not sliding, jumping or falling.  
Make sure they are not firing or possibly even have their guns away. 
Adding a min health cap would probably help, i.e no auto save below 10%
Possibly doing it when you enter a new room, its not a 100% "a portal" but it is mostly when you enter a new area, or "after you enter a new area" in the case where you start/enter with a slide. This gives a nice logical break. I also doubt there is a room cross that leads to instant death, unless you are falling.

Collecting an item is probably also a good spot, i.e the save takes you back to the last cog, the last statue, the last med pack. Those are mostly in safe spots. 

I also feel that having 2~3 last auto saves would be better since our out saves are not "designer placed" allowing them to go back further if they need to, would be a good idea.

**Comment by Richard-L** [comment]
Consider quicksave/load instead, as in no passport and nothing, just a frozen frame and an OSD confirmation. Basically a button people can press ahead of most jumps, and another to reload. Almost acts as a quick rewind that way.

OpenLara does this.

**Comment by stohrendorf** [comment]
Keep in mind that it somewhat contradicts the original gameplay of being somewhat of a "survival game", i.e. the original game was meant to be "hard" - if you made a mistake, you _felt_ you made a mistake. If you add auto-save, it's making tomb raider a lot less stressful, and basically would make it mostly harmless, as you could load a save automatically made a few seconds ago whenever you make a mistake. I understand it feels frustrating when you forgot to save, but IMO it's a fundamental feature of the gameplay. This is why I never wanted to implement that for CE, although I thought about it for quite a while. I don't want to stop you from implementing it, just throwing in some thoughts I had, possibly giving you another perspective. (Not to mention the complexity about _when_ autosave should occur, as @oziphantom already pointed out.)

**Comment by oziphantom** [comment]
sure if playing the game "for the game" you don't want auto save. However when streaming. Streaming is a lot harder, in Fire Emblem we joke you only have 30% power to play with, hence when you stream you make a lot of stupid moves. This is because focusing on chatting, reading chat, being `entertaining` takes a lot of thought, and remember to open up and save all the time is a lot of effort on top of it. Also watching people redo things again is not "entertaining". If you are streaming it then you probably have already played it and everybody watching probably has as well, so its not really ruining the game for anybody. As an optional "for streamers" it's a very nice to have.

**Comment by rr-** [comment]
I agree that this would stray off the OG experience, but as such, it would be optional like most of our features are. And like Ozi said I was mostly thinking about Twitch, where it's a quite common thing to see people getting frustrated with not making a save recently enough (including the chat, not just the streamers).

**Comment by Pia86** [comment]
How about making an autosave at the location of the save crystals in this mode?
